In [2]:
# import all necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

#### Question 1
     What's the version of Pandas that you installed?

In [3]:
pd.__version__

'1.4.2'

In [4]:
# load dataset
housing_data = pd.read_csv(r"C:\Users\User\Documents\Tech\ML Zoomcamp\alexeygrigorev_datasets_master_housing.csv")

In [6]:
housing_data

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

##### Question 2
    How many columns are in the dataset?

In [8]:
housing_data.shape

(20640, 10)

There are 20640 rows and 10 columns

#### Question 3

    Which columns in the dataset have missing values?

In [9]:
housing_data.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

Column with mssing data is total_bedrooms

#### Question 4
    How many unique values does the ocean_proximity column have?

In [10]:
housing_data.nunique()

longitude               844
latitude                862
housing_median_age       52
total_rooms            5926
total_bedrooms         1923
population             3888
households             1815
median_income         12928
median_house_value     3842
ocean_proximity           5
dtype: int64

ocean_proximity has 5 unique values

#### Question 5
    What's the average value of the median_house_value for the houses located near the bay?

In [15]:
average_value_near_bay = housing_data[housing_data['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean()
print(average_value_near_bay)


259212.31179039303


#### Question 6 
    a. Calculate the average of total_bedrooms column in the dataset.
    b. Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
    c. Now, calculate the average of total_bedrooms again.
    d. Has it changed?

In [16]:
# a
average_total_bedrooms = housing_data['total_bedrooms'].mean()
print(average_total_bedrooms)


537.8705525375618


In [17]:
#b. 
# Assuming you have a DataFrame called 'housing_data'
# Calculate the mean of the 'total_bedrooms' column
average_total_bedrooms = housing_data['total_bedrooms'].mean()

# Fill missing values in 'total_bedrooms' with the calculated mean
housing_data['total_bedrooms'].fillna(average_total_bedrooms, inplace=True)


In [18]:
# Calculate the mean of 'total_bedrooms' before filling missing values
mean_before_fill = housing_data['total_bedrooms'].mean()

# Fill missing values in 'total_bedrooms' with the calculated mean
housing_data['total_bedrooms'].fillna(mean_before_fill, inplace=True)

# Calculate the mean of 'total_bedrooms' after filling missing values
mean_after_fill = housing_data['total_bedrooms'].mean()

# Compare the mean values before and after filling
if mean_before_fill == mean_after_fill:
    print("Mean value did not change after filling missing values.")
else:
    print("Mean value changed after filling missing values.")


Mean value did not change after filling missing values.


#### Question 7
    a. Select all the options located on islands.
    b. Select only columns housing_median_age, total_rooms, total_bedrooms.
    c. Get the underlying NumPy array. Let's call it X.
    d. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
    e. Compute the inverse of XTX.
    d. Create an array y with values [950, 1300, 800, 1000, 1300].
    e. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
    f. What's the value of the last element of w?

In [31]:
# a.
options_on_islands = housing_data[housing_data["ocean_proximity"] == "NEAR BAY"]

In [22]:
#b.
selected_columns = ['housing_median_age', 'total_rooms', 'total_bedrooms']
subset_housing_data = housing_data[selected_columns]
selected_columns

['housing_median_age', 'total_rooms', 'total_bedrooms']

In [30]:
# c.
X = housing_data.values
